### PR TITLE
ci: validate the message of each new commit as part of the CI linting

### PR DESCRIPTION
**Pull request by IgorMinar**

### CODE REVIEW

File: tools/validate-commit-message/index.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  const type = match[1];
+  if (config['types'].indexOf(type) === -1) {
+    error(`${type} is not an allowed type.\n => TYPES: ${config['types'].join('|')}`, commitMessage);
+    return false;
+  }
+
+  const scope = match[2];
+  if (scope && scope.length > 0 && config['scopes'].indexOf(scope) === -1) {
```
Comment (vicb) [inline tools/validate-commit-message/index.js]: `scope.length > 0` useless

File: tools/validate-commit-message/index.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env node
+
+/**
+ * GIT COMMIT-MSG hook for validating commit message.
+ * See https://docs.google.com/document/d/1rk04jEuGfk9kYzfqCuOlPTSJw3hEDZJTBN5E5f1SALo
+ *
+ * Installation:
+ * >> cd <repo>
+ * >> ln -s ../../lib/validate-commit-msg.js .git/hooks/commit-msg
```
Comment (vicb) [inline tools/validate-commit-message/index.js]: If this is intended to be a hook, we should split the headline vs the body (we do not want to restrict the body length)

File: gulpfile.js [hunk 3]
```diff
@@ -155,6 +155,42 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
...
+  const childProcess = require('child_process');
+
+  childProcess.exec('git log --reverse --oneline HEAD ^upstream/master', (error, stdout, stderr) => {
+    if (error) {
+      console.log(stderr);
+      process.exit(1);
+    }
+
+    let someCommitsInvalid = false;
+    let commitsByLine = stdout.trim().split(/\n/);
```
Comment (vicb) [inline gulpfile.js]: trim ?

File: gulpfile.js [hunk 4]
```diff
@@ -155,6 +155,42 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
       .pipe(tslint.report({emitError: true}));
 });
 
+gulp.task('tools:validate-commit-message', () => {
+  const validateCommitMessage = require('./tools/validate-commit-message');
+  const childProcess = require('child_process');
+
+  childProcess.exec('git log --reverse --oneline HEAD ^upstream/master', (error, stdout, stderr) => {
```
Comment (vicb) [inline gulpfile.js]: `git log --reverse --oneline HEAD ^FETCH_HEAD`
actually `git log --reverse --oneline HEAD ^master` is probably good ?

File: gulpfile.js [hunk 5]
```diff
@@ -155,6 +155,42 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
...
+    }
+
+    let someCommitsInvalid = false;
+    let commitsByLine = stdout.trim().split(/\n/);
+
+    console.log(`Examining ${commitsByLine.length} commits between HEAD and master`);
+
+    if (commitsByLine.length == 0) {
+      console.log('There are zero new commits between this HEAD and master');
+      process.exit(1);
```
Comment (vicb) [inline gulpfile.js]: 1 ?

File: gulpfile.js [hunk 6]
```diff
@@ -155,6 +155,42 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
...
+    }
+
+    commitsByLine.forEach(line => {
+      // the oneline format per line is: "SHA MESSAGE"
+      const firstSpace = line.indexOf(' ');
```
Comment (vicb) [inline gulpfile.js]: do not request the SHA if we do not use it ? `git log --reverse HEAD ^master --format=%s`

File: tools/validate-commit-message/index.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const configPath = path.resolve(__dirname, './commit-message.json');
+const config = JSON.parse(fs.readFileSync(configPath, "utf8"));
+const PATTERN = /^(\w+)(?:\(([\w\$\.\*/-]*)\))?\: (.+)$/;
+
+module.exports = function(commitMessage) {
```
Comment (vicb) [inline tools/validate-commit-message/index.js]: commitMessage -> subject

File: tools/validate-commit-message/index.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+ * >> chmod +x .git/hooks/commit-msg
+ */
+
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const configPath = path.resolve(__dirname, './commit-message.json');
+const config = JSON.parse(fs.readFileSync(configPath, "utf8"));
+const PATTERN = /^(\w+)(?:\(([\w\$\.\*/-]*)\))?\: (.+)$/;
```
Comment (vicb) [inline tools/validate-commit-message/index.js]: `[\w\$\.\*/-]*` -> `[\w-]+` ?

File: tools/validate-commit-message/index.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+  const match = PATTERN.exec(commitMessage);
+  if (!match) {
+    error(`The commit message does not match the format of "<type>(<scope>): <subject>"`, commitMessage);
+    return false;
+  }
+
+  const type = match[1];
+  if (config['types'].indexOf(type) === -1) {
+    error(`${type} is not an allowed type.\n => TYPES: ${config['types'].join('|')}`, commitMessage);
```
Comment (vicb) [inline tools/validate-commit-message/index.js]: join(', ') for readability ?

File: tools/validate-commit-message/index.js (hunk 8 above)
Comment (vicb) [inline tools/validate-commit-message/index.js]: IIUC currently `type(): ...` would be a false positive

**Comment by IgorMinar** [comment]
overriding CLA because @matsko signed it and I'm just taking over his PR with his permission.

File: tools/validate-commit-message/index.js (hunk 1 above)
Comment (vicb) [inline tools/validate-commit-message/index.js]: indexOf === -1 -> .includes ?

**Comment by IgorMinar** [comment]
@vicb I've cleaned everything up, added docs and tests. it should be good to go.

in the end I didn't use `git log --reverse --oneline HEAD ^FETCH_HEAD` because I think that's a bit unpredictable on local machine where what you last fetched might not be the last thing that's in upstream or origin. diffing against origin seems to work just fine.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -191,21 +191,46 @@ If the commit reverts a previous commit, it should begin with `revert: `, follow
...
+* **tsc-wrapped**
+
+There is currently few exception to the "use package name" rule:
+
+* **packaging**: used for changes that change the npm package layout in all of our packages, e.g. public path changes, package.json changes done to all packages, d.ts file/format changes, changes to bundles, etc.
+* **changelog**: used for updating the release notes in CHANGELOG.md
+* none/empty string: useful for `style`, `test` and `refactor` changes that are done across all packages (e.g. `style: add missing semicolons`)
+
+
+Packagigng
```
Comment (gkalpak) [inline CONTRIBUTING.md]: Typo?

File: gulpfile.js [hunk 11]
```diff
@@ -155,6 +155,42 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
...
+
+        if (commitsByLine.length == 0) {
+          console.log('There are zero new commits between this HEAD and master');
+        }
+
+        commitsByLine.forEach(commitMessage => {
+          if (!validateCommitMessage(commitMessage)) {
+            someCommitsInvalid = true;
+          }
+        });
```
Comment (gkalpak) [inline gulpfile.js]: `someCommitsInvalid = !commitsByLine.every(validateCommitMessage)` is shorter, more readable and efficient 😛

File: CONTRIBUTING.md (hunk 10 above)
Comment (IgorMinar) [inline CONTRIBUTING.md]: yes. fixed

File: gulpfile.js (hunk 11 above)
Comment (IgorMinar) [inline gulpfile.js]: done. thanks

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -191,21 +191,46 @@ If the commit reverts a previous commit, it should begin with `revert: `, follow
...
-The scope could be anything specifying place of the commit change. For example
-`Compiler`, `ElementInjector`, etc.
+The scope should be the name of the npm package affected (as perceived by person reading changelog generated from commit messages.
+
+The following is the list of supported scopes:
+
+* **common**
+* **compiler**
+* **compiler-cli**
+* **core**
```
Comment (vicb) [inline CONTRIBUTING.md]: What about `i18n` ? Should we allow `compiler/i18n` ? 
Not blocking the PR but let's discuss this

File: gulpfile.js [hunk 13]
```diff
@@ -155,6 +155,40 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
...
 });
 
+gulp.task('validate-commit-messages', () => {
+  const validateCommitMessage = require('./tools/validate-commit-message');
+  const childProcess = require('child_process');
+
+  // We need to fetch origin explicitly because it might be stale.
+  // I couldn't find a reliable way to do this without fetch.
+  childProcess.exec(
+      'git fetch origin master && git log --reverse --format=%s HEAD ^origin/master',
```
Comment (vicb) [inline gulpfile.js]: couldn't we use the local `master`:
- should be the same as `origin/master` on the CI,
-should be at least the fork point on your local dev

File: gulpfile.js [hunk 14]
```diff
@@ -155,6 +155,40 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
...
+          console.log(stderr);
+          process.exit(1);
+        }
+
+        let someCommitsInvalid = false;
+        let commitsByLine = stdout.trim().split(/\n/);
+
+        console.log(`Examining ${commitsByLine.length} commits between HEAD and master`);
+
+        if (commitsByLine.length == 0) {
```
Comment (vicb) [inline gulpfile.js]:
```
if (0) exit(0)
else
   ...
```

File: CONTRIBUTING.md (hunk 10 above)
Comment (gkalpak) [inline CONTRIBUTING.md]: I should have been clearer: I meant that the whole word should probably be removed.

File: tools/validate-commit-message/validate-commit-message.js [hunk 15]
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env node
+
+/**
+ * GIT COMMIT-MSG hook for validating commit message.
+ * See https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines
+ *
+ * Installation:
+ * >> cd <repo>
+ * >> ln -s ../../tools/validate-commit-message/validate-commit-message.js .git/hooks/commit-msg
```
Comment (vicb) [inline tools/validate-commit-message/validate-commit-message.js]: not addressed

File: tools/validate-commit-message/validate-commit-message.js [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+ * Note: this script was originally written by Vojta for AngularJS :-)
+ */
+
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const configPath = path.resolve(__dirname, './commit-message.json');
+const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
+const PATTERN = /^(revert\: )?(\w+)(?:\(([\w\$\.\*/-]*)\))?\: (.+)$/;
```
Comment (vicb) [inline tools/validate-commit-message/validate-commit-message.js]: previous comment not addressed

File: tools/validate-commit-message/validate-commit-message.js [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+
+module.exports = function(commitMessage) {
+  if (commitMessage.length > config['maxLength']) {
+    error(`The commit message is longer than ${config['maxLength']} characters`, commitMessage);
+    return false;
+  }
+
+  const match = PATTERN.exec(commitMessage);
+  if (!match || match[2] === 'revert') {
```
Comment (vicb) [inline tools/validate-commit-message/validate-commit-message.js]: why is revert special here ?

File: tools/validate-commit-message/validate-commit-message.spec.js [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+      expect(validateMessage('style(http): something')).toBe(VALID);
+      expect(validateMessage('refactor(platform-webworker): something')).toBe(VALID);
+      expect(validateMessage('test(language-service): something')).toBe(VALID);
+      expect(validateMessage('test(packaging): something')).toBe(VALID);
+      expect(errors).toEqual([]);
+    });
+
+
+    it('should fail when scope is invalid', function() {
+      expect(validateMessage('fix(Compiler): something')).toBe(INVALID);
```
Comment (vicb) [inline tools/validate-commit-message/validate-commit-message.spec.js]: to we want to be case insensitive ?

File: tools/validate-commit-message/validate-commit-message.js [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    error(
+        `${type} is not an allowed type.\n => TYPES: ${config['types'].join('|')}`, commitMessage);
+    return false;
+  }
+
+  const scope = match[3];
+
+  if (scope && !config['scopes'].includes(scope)) {
+    error(
+        `"${scope}" inside of "${type}(...)" is not an allowed scope.\n => SCOPES: ${config['scopes'].join('|')}`,
```
Comment (vicb) [inline tools/validate-commit-message/validate-commit-message.js]: remove or rephrase `inside of "${type}(...)"` when looking at the error msgs in the tests, what I read is that I can use whatever scope but not inside the current type vs the scope is invalid.

File: tools/validate-commit-message/validate-commit-message.js (hunk 15 above)
Comment (IgorMinar) [inline tools/validate-commit-message/validate-commit-message.js]: hmm.. what was the issue with this? I'll look up your old comment.

File: tools/validate-commit-message/validate-commit-message.js (hunk 17 above)
Comment (IgorMinar) [inline tools/validate-commit-message/validate-commit-message.js]: it's always been this way. we just never enforced it. the revert should contain the whole original message so it's not a special type. it's more like a prefix for the usual message format.

File: tools/validate-commit-message/validate-commit-message.spec.js (hunk 18 above)
Comment (IgorMinar) [inline tools/validate-commit-message/validate-commit-message.spec.js]: the packages are all lowercase. enforcing the case is a good thing for consistency.

File: CONTRIBUTING.md (hunk 12 above)
Comment (IgorMinar) [inline CONTRIBUTING.md]: I'm implementing the current conventions that match the current changelog output. we can discuss can discuss bigger changes to this conventions in a separate commit.

File: gulpfile.js (hunk 13 above)
Comment (IgorMinar) [inline gulpfile.js]: I tried but on Circle CI the master is almost always stale for some reason. this was the only way I could get reliably results.

File: tools/validate-commit-message/validate-commit-message.js (hunk 15 above)
Comment (IgorMinar) [inline tools/validate-commit-message/validate-commit-message.js]: > If this is intended to be a hook, we should split the headline vs the body (we do not want to restrict the body length)

hmm. you are right. originally when Vojta wrote it was meant to be a commit hook, but with all the changes it turned into something else.

I'll change the code to check just the first line.

Comment (IgorMinar) [inline tools/validate-commit-message/validate-commit-message.js]: actually that would be a more complicated change. I'll just drop the comment.

File: tools/validate-commit-message/index.js (hunk 8 above)
Comment (IgorMinar) [inline tools/validate-commit-message/index.js]: fixed. good catch

File: tools/validate-commit-message/index.js (hunk 9 above)
Comment (IgorMinar) [inline tools/validate-commit-message/index.js]: done

**Comment by vicb** [comment]
LGTM, needs a green CI (cleanup)

**Pull request merged**